server/src/main/java/io/spine/server/projection/ProjectionRepository.java

```diff
@@ -338,40 +336,148 @@ public final boolean canDispatch(EventEnvelope event) {
+    /**
+     * Repeats the dispatching of the events from the event log to the all the entities managed
+     * by this repository. The events are read from the specified point in time.
+     *
+     * <p>At the beginning of the process the state of all of the entities is set to the default.
+     *
+     * <p>The events are dispatched according to the actual event routing schema. So a historical
+     * event may be dispatched to a different set of targets, comparing to those to which it was
+     * dispatched once emitted originally.
+     *
+     * <p>The subscriptions to the entity state updates (i.e.
+     * {@link io.spine.system.server.event.EntityStateChanged EntityStateChanged}) events are not
+     * supported in the catch-up.
+     *
+     * @param since
+     *         point in the past, since which the catch-up should be performed
+     * @throws CatchUpAlreadyStartedException
+     *         if another catch-up for the same entity type is already in progress
+     */
+    public void catchUpAll(Timestamp since) throws CatchUpAlreadyStartedException {
+        catchUp(since, null);
+    }
+
+    /**
+     * Sends the event to the inboxes of the catching-up projection instances.
+     *
+     * <p>Allows to restrict the target entities by identifiers. In this case, the event is
+     * routed as per the repository routing schema, and the obtained set of the identifiers
+     * is narrowed down to the restricted targets.
+     *
+     * <p>Such a setting allows to catch up only the selected targets.
+     *
+     * <p>This API method also supports sending the special {@link CatchUpSignal}s
+     * to the projection. They regulate the lifecycle of the catch-up and are handled by
+     * the {@link CatchUpEndpoint} exposed by this repository.
+     *
+     * <p>Please note that the {@code CatchUpSignal}s are dispatched to the selected targets
+     * only and cannot be dispatched to all of the repository instances. The reason is that
+     * handling of {@code CatchUpSignal}s may affect the lifecycle state of the projection
+     * instances. E.g. the callee must know to what targets he is sending the "delete state" signal.
+     *
+     * @param event
+     *         the event to dispatch
+     * @param restrictToIds
+     *         optional set of the target identifiers to which the dispatching must be restricted;
+     *         if {@code null}, no restriction are applied and the event should be dispatched as per
+     *         the routing schema
+     * @return the set of the entity identifiers, which actually received the dispatched event
+     * @see CatchUpEndpoint
+     */
+    private Set<I> sendToCatchingUp(Event event, @Nullable Set<I> restrictToIds) {
+        EventEnvelope envelope = EventEnvelope.of(event);
+        Set<I> catchUpTargets;
+        if (envelope.message() instanceof CatchUpSignal) {
+            catchUpTargets = restrictToIds == null
+                             ? ImmutableSet.copyOf(index())
```

Comment:
This operation is potentially big and long (assuming we go for the `index()` to the storage). Can we run it once per repository under the catch up operation?

Reply:
The idea here is to have the most recent `index()`, so we shouldn't cache it.

In the real use cases, it is going to be used only if a catch-up for all the repository entities is requested.  If this repository is a new one and requires catching up, the index is going to be empty. And if the repository entities are all messed up and require the complete re-catch-up,
it will and should take long. So the expectations are managed.